### PR TITLE
Use cmake to build wasmtime-c-api

### LINKED ISSUE #8890: `#include <wasmtime/conf.h>` fails since `conf.h` is not generated
When we try to bump to wasmtime v22.0.0 in https://github.com/tree-sitter/tree-sitter/pull/3428 we get the following error:

```
cargo:warning=/home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmtime-c-api-impl-22.0.0/include/wasi.h:12:10: fatal error: wasmtime/conf.h: No such file or directory
  cargo:warning=   12 | #include <wasmtime/conf.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~
```

When I checked this locally it seems that `conf.h` is not being generated when building the c-api via a crate as I could not find it in `cargo:warning=/home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmtime-c-api-impl-22.0.0/include/`, which is where I assumed it would be as that's where the headers are looking for it.

### PR DESCRIPTION
**Pull request by CGamesPlay**
This resolves #8890, and allows a rust project to use the `wasmtime-c-api` crate again.

In #8642, the build process was changed but the rust build script was not changed to match. In addition, because the CMake configuration invokes cargo, it isn't possible for cargo to then invoke CMake (cylic dependency). To resolve this, we introduce a new CMake target that only produces the necessary files for the build to succeed, and then allow cargo to build the API as normal.

I've tested that this works in my local copy of tree-sitter, which depends on the `wasmtime-c-api-impl` crate.

### CODE REVIEW

**Review by alexcrichton — COMMENTED** [review]
Thanks! Do you know if it's possible to have an install target which only installs header files? That way there wouldn't need to be two includes and only one would be necessary.

Additionally can you forward along the `-D` flags for the features of Wasmtime? Otherwise the `conf.h` could report the wrong set of supported features.

**Comment by CGamesPlay** [comment]
Sure, we just need to use `COMPONENT Headers`, see [docs](https://cmake.org/cmake/help/latest/command/install.html). I think only having one header would be ideal, especially if they are in different directories.

Are you suggesting that instead of just generating the header, I install all the headers to the cargo output directory, and then report that as the sole include directory? That makes sense, I just want to confirm.

**Comment by CGamesPlay** [comment]
Updated commits properly configure the headers, and install all headers to the right place, so we can get rid of the second metadata variable.

**Review by alexcrichton — APPROVED** [review]
Thanks!

**Comment by clason** [comment]
Can this be backported to release-23, too?

**Comment by alexcrichton** [comment]
I think that would be reasonable yeah, just needs to land on `main` first and then the backport is a cherry-pick + new PR. 

I believe CI is failing here on MinGW for cmake-related reasons? (compiler detection not working presumbably because `cl.exe` is used when mingw flags are passed)

**Comment by maxbrunsfeld** [comment]
Is there any way we could avoid adding this `cmake` build dependency to `wasmtime-c-api-impl`?

Could the `build.rs` just generate `conf.h`, instead of using cmake to do it? It looks like that `build.rs` file already has to explicitly list out a bunch of cargo features and convert them to C macro names. Could it just write that to a `conf.h` file directly, instead of passing them as arguments to cmake?

Maybe I'm missing something, and the CMake build is doing something too significant to duplicate, but if not, it would be great to avoid the build-time dependency.

**Comment by alexcrichton** [comment]
Is the `cmake` dependency too heavy for end users to arrange? I do think it's nice to have the same logic not duplicated. It's already broken once now so keeping it synchronized seems like the best bet.

**Comment by maxbrunsfeld** [comment]
I just anticipate that  consumers of the `tree-sitter` crate won't necessarily have `cmake` installed on their CI / dev machines, if they are mainly set up for Rust development.

Just to double check, I think that if we wanted to avoid requiring Cmake, we'd change this:

```rust
    let dir = PathBuf::from(env::var_os("CARGO_MANIFEST_DIR").unwrap());
    let mut config = cmake::Config::new(&dir);
    config
        .define("WASMTIME_DISABLE_ALL_FEATURES", "ON")
        .always_configure(true)
        .build_target("headers");
    for f in FEATURES {
        if env::var_os(format!("CARGO_FEATURE_{}", f)).is_some() {
            config.define(format!("WASMTIME_FEATURE_{}", f), "ON");
        }
    }
    let dst = config.build();
```

into this (or something similar). 
```rust
    let prefix = "#ifndef WASMTIME_CONF_H\n#define WASMTIME_CONF_H\n";
    let suffix = "#endif\n";

    let mut header = prefix.to_string();
    for f in FEATURES {
        if std::env::var_os(format!("CARGO_FEATURE_{}", f)).is_some() {
            writeln!(&mut header, "WASMTIME_FEATURE_{}", f);
        }
    }
    header += suffix;
```

And you wouldn't need to add the `headers` target to the cmake file. It doesn't seem any more brittle to me. Either way, the build script needs to know the set of features, and how they map to C macro names.

**Comment by alexcrichton** [comment]
The brittle part is that it's a duplication of copying over headers and there is no longer a single source of truth for how headers are created. Additionally there are no tests in this repository testing the header as-generated by the build script and without that there's also no tests that the duplicated logic for the header file generation is correct. What you're describing is correct as-is today but that may not hold true indefinitely into the future.

**Comment by maxbrunsfeld** [comment]
Oh, I would have thought that if there are tests for `wasmtime-c-api`, then they would cover the correctness of the `wasmtime-c-api-impl` build script. Is there a tested code path that doesn't run this build script?

**Comment by CGamesPlay** [comment]
Is there are test that mingw compilation is working at all presently? This invocation of cmake looks correct, but it detects MSVC anyways:

```
    running: "cmake" "D:\\a\\wasmtime\\wasmtime\\crates\\c-api"
        "-G" "MSYS Makefiles"
        "-DWASMTIME_DISABLE_ALL_FEATURES=ON"
        "-DWASMTIME_FEATURE_DISABLE_LOGGING=ON"
        "-DCMAKE_SYSTEM_NAME=Windows"
        "-DCMAKE_SYSTEM_PROCESSOR=AMD64"
        "-DCMAKE_INSTALL_PREFIX=D:\\a\\wasmtime\\wasmtime\\target\\x86_64-pc-windows-gnu\\release\\build\\wasmtime-c-api-impl-dfbdb5120f123319\\out"
        "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -m64"
        "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -m64"
        "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -m64"
        "-DCMAKE_BUILD_TYPE=MinSizeRel"
    -- The C compiler identification is MSVC 19.40.33812.0
```

**Comment by CGamesPlay** [comment]
I've added a workflow change to include the commands that I suspect fix the issue.

**Comment by alexcrichton** [comment]
I've added an extra commit to this PR which runs the full tests suite on this PR instead of waiting for the merge queue, and it looks like there may still be some CI issues

**Comment by CGamesPlay** [comment]
The one remaining CI failure is a network failure. Changes made to make CI happy:

- Force CC and CXX in mingw. Turns out the other mingw build configurations don't build the C API anyways.
- Use `cargo package --no-verify` on the C API since it is hard-coded to create an `artifact` directory in the source directory and it isn't possible to package that directory because cargo is hard-coded to exclude any directory with `Cargo.toml` in it.
  - By the way, the `--no-verify` on `wasi-nn` seems unnecessary (it worked fine on my machine).

File: crates/c-api/CMakeLists.txt
Comment (alexcrichton) [inline crates/c-api/CMakeLists.txt:104]: Should the `BINARY_DIR` configuration be dropped entirely? That should probably be somewhere within `CMAKE_CURRENT_BINARY_DIR` if it's still being used to avoid modifying the source tree.

**Review by alexcrichton — COMMENTED** [review]
Thanks! Given the possible fragility here with packaging and how you had to modify `include`, could verification via `publish.rs` be made to work? Basically could this be updated to undo the addition of `--no-verify`?

As for wasi-nn, if you'd like I think it's reasonable to try dropping that here too (and verify it), but it's ok to defer that to a different PR as well.

File: .github/workflows/main.yml
Comment (alexcrichton) [inline .github/workflows/main.yml:1093]: Mind leaving a comment as to why these steps are necessary? (this doesn't seem to directly follow from rust-lang/rust#112368 by my read

**Comment by CGamesPlay** [comment]
> Given the possible fragility here with packaging and how you had to modify include, could verification via publish.rs be made to work? Basically could this be updated to undo the addition of --no-verify?

Short answer: no. Longer answer: yes, but I probably can't do it in this PR. Any change I make here will at best be "works on my machine", and will likely involve knowing the details of all supported configurations and what went into https://github.com/bytecodealliance/wasmtime/pull/8642. Plus, I'm not a core developer of this project, I'm trying to get access to the 1-line method I added in https://github.com/bytecodealliance/wasmtime/pull/8907 so that I can update tree-sitter so that I can use wasm modules in my project. Basically, I'm already 2 projects deep in a rabbit hole 🙂

Hopefully you'll appreciate that I've taken the rust c-api crate from definitely-not-working to working-but-not-thoroughly-tested, and can add a follow-up task to the backlog to improve the testing.

**Comment by alexcrichton** [comment]
I'm losing steam on this and I don't know how best to proceed. Running a full `cmake` configure step has issues with needing extra CI configuration and new build dependencies (not just `cmake` but the build system like make/ninja as well). It additionally resurfaced android errors in https://github.com/bytecodealliance/wasmtime/actions/runs/10217416984/job/28271193012 which I don't know what's going on.

I tried an alternate approach of installing headers with a `*.cmake` script to use `cmake -P` instead. That skips the whole configure step and only requires a single `cmake` invocation and should be pretty portable and more lightweight. That is causing all sorts of issues of having to install headers in a few places as part of the build for things like docs and examples. I don't know how to best handle that.

Overall I don't know the best way forward on this. It might be to stop relying on cmake for features and use a rust-based script that can be shared between the normal cmake build and the build script. Unsure.

**Comment by CGamesPlay** [comment]
I’m not at my computer right now but about the Android failure, one thing I noticed is that the NDK is v27 on -min and v25 on the regular one, with a different rust toolchain and a different runner image version. I set `CMAKE_MAKE_PROGRAM` to `make` and I got a successful run, but I am suspicious that its actually a nondeterministic failure.

File: Cargo.toml
Comment (clason) [inline Cargo.toml:349]:
```suggestion
# When adding or removing a feature, make sure to keep the C API in sync by
```

File: crates/c-api/include/wasmtime/async.h
Comment (clason) [inline crates/c-api/include/wasmtime/async.h:14]: unrelated changes?

File: crates/c-api/include/wasmtime/conf.h.in
Comment (clason) [inline crates/c-api/include/wasmtime/conf.h.in:27]: If you move this comment _above_ the list (here and elsewhere), then you can avoid the duplicate nonce (less noise when `grep`ping).

**Review by alexcrichton — APPROVED** [review]
Ah thanks for debugging that! This looks good to me yeah, so I'm going to go ahead and merge. I'm going to follow up with some responses to @clason as well and there can always be follow-ups too to handle things.

File: crates/c-api/include/wasmtime/conf.h.in
Comment (alexcrichton) [reply crates/c-api/include/wasmtime/conf.h.in:27]: Personally I like having this at the end because features are often added at the end and a comment only at the top might get missed. The amount of duplication here though is unfortunate across the codebase but that would probably be best fixed with a "tidy" script or something like that rather than relying on us humans to handle everything.

File: crates/c-api/include/wasmtime/async.h
Comment (alexcrichton) [reply crates/c-api/include/wasmtime/async.h:14]: I found this was necessary at one stage of testing to appease doxygen. That being said I barely understanding doxygen and we could very well be holding doxygen wrong where this should work when it doesn't.

File: crates/c-api/include/wasmtime/conf.h.in
Comment (clason) [reply crates/c-api/include/wasmtime/conf.h.in:27]: I just mean as a trivial way to halve the number of hits: no need to have the nonce both before _and_ after the list; once is enough (no matter where).

**Comment by clason** [comment]
And also backport (24.0 is probably good enough) ;)

**Comment by CGamesPlay** [comment]
I'd definitely love to get this in 24 if possible. Link to @clason's backport PRs:

- #9101
- #9102

**Pull request merged**
